### PR TITLE
Add the possibility to change the type of an included structural element

### LINKED ISSUE #3372: Change the type of a structural element
Once a structural element has been created, it can have children and assigned pages. Its type cannot be changed any more at the moment.

**Goal:** It should be possible to open a dialog via the context menu to change the type of the structural element. You can choose ~~between~~ only types that are permissible for the given parent and accept all existing children ~~or all possible types of the parent, whereby dirty children are possible as a result.~~ 

Example:
Example dialogue

### PR DESCRIPTION
**Pull request by matthias-ronge**
Fixes #3372

### CODE REVIEW

**Comment by matthias-ronge** [comment]
I don't understand the error Codacy is accusing. I wrote the line in Javascript the same way as the line above it, and it works, too. It cannot be that the function is "undefined".

**Comment by oliver-stoehr** [comment]
> I don't understand the error Codacy is accusing. I wrote the line in Javascript the same way as the line above it, and it works, too. It cannot be that the function is "undefined".

As `PF` is not defined in `resize.js` Codacy cannot find the definition. You must declare it as global variable to suppress this error (See `metadata_editor.js:11`).

File: Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/ChangeDocStrucTypeDialog.java [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+/**
+ * Backing bean for the change doc struc type dialog of the metadata editor.
+ */
+public class ChangeDocStrucTypeDialog {
+    private static final Logger logger = LogManager.getLogger(ChangeDocStrucTypeDialog.class);
+
+    private final DataEditorForm dataEditor;
+    private final List<SelectItem> docStructEditTypeSelectionItems = new ArrayList<>();
+    private String docStructEditTypeSelectionSelectedItem;
```
Comment (Kathrin-Huber) [inline Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/ChangeDocStrucTypeDialog.java:44]:
```suggestion
    private String docStructType;
```

File: Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/ChangeDocStrucTypeDialog.java [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+import org.kitodo.production.services.ServiceManager;
+
+/**
+ * Backing bean for the change doc struc type dialog of the metadata editor.
+ */
+public class ChangeDocStrucTypeDialog {
+    private static final Logger logger = LogManager.getLogger(ChangeDocStrucTypeDialog.class);
+
+    private final DataEditorForm dataEditor;
+    private final List<SelectItem> docStructEditTypeSelectionItems = new ArrayList<>();
```
Comment (Kathrin-Huber) [inline Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/ChangeDocStrucTypeDialog.java:43]:
```suggestion
    private final List<SelectItem> docStructTypes = new ArrayList<>();
```

File: Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/ChangeDocStrucTypeDialog.java [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+            for (Entry<String, String> typeOption : possibleTypes.entrySet()) {
+                docStructEditTypeSelectionItems.add(new SelectItem(typeOption.getKey(), typeOption.getValue()));
+            }
+            docStructEditTypeSelectionSelectedItem = selectedStructure.getType();
+        } catch (IllegalStateException | IOException e) {
+            Helper.setErrorMessage(e.getLocalizedMessage(), logger, e);
+        }
+    }
+
+    private IncludedStructuralElement getTheCurrentlySelectedIncludedStructuralElement() {
```
Comment (Kathrin-Huber) [inline Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/ChangeDocStrucTypeDialog.java:111]:
```suggestion
    private IncludedStructuralElement getSelectedIncludedStructuralElement() {
```

File: Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/ChangeDocStrucTypeDialog.java [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        if (dataEditor.getSelectedStructure().isPresent()) {
+            return dataEditor.getSelectedStructure().get();
+        } else {
+            throw new IllegalStateException("could not retrieve selected structure");
+        }
+    }
+
+    private Map<String, String> findAllTypesThisIncludedStructuralElementCanHave(
```
Comment (Kathrin-Huber) [inline Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/ChangeDocStrucTypeDialog.java:119]:
```suggestion
    private Map<String, String> findAllPossibleTypes(
```

File: Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/ChangeDocStrucTypeDialog.java [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+            ((LinkedList<IncludedStructuralElement>) parentHierarchyPath).getLast().getType());
+    }
+
+    private Map<String, String> getAllowedSubstructuralElements(String parentType) {
+        StructuralElementViewInterface parentView = dataEditor.getRuleset().getStructuralElementView(parentType,
+            dataEditor.getAcquisitionStage(), dataEditor.getPriorityList());
+        return parentView.getAllowedSubstructuralElements();
+    }
+
+    private void onlyAllowTypesThatSupportAllGivenIncludedStructuralChildElements(
```
Comment (Kathrin-Huber) [inline Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/ChangeDocStrucTypeDialog.java:167]:
```suggestion
    private void restrictTypesToChildElements(
```

**Comment by matthias-ronge** [comment]
Codacy has another problem:

> A function with a name starting with an uppercase letter should only be used as a constructor.

I assume I cannot change that because it is built-in?

**Comment by oliver-stoehr** [comment]
Since `PF` is a function by PrimeFaces we cannot change the name.
@Kathrin-Huber can you rerun Codacy ignoring `PF`?

**Pull request merged**